library/core/src/array/mod.rs

```diff
@@ -879,7 +879,7 @@ where
         }
     }
 
-    let mut array = MaybeUninit::uninit_array::<N>();
+    let mut array = MaybeUninit::uninit().transpose();
```

Comment:
Hmmm, yeah I was wondering about this too. My reasoning for this being ok was that if the compiler manages to infer the types then I don't need to think about them. I'm guessing this is flawed because if you have no constraints *at all*, then the compiler will happily give you an i32. An interesting question arises: are integers and floats the only two places where the compiler will ever pull a type out of its :peach: if there are no constraints? If so, my gut feeling is that we should get rid of that behavior (separate discussion, I know) since not having :peach: types would let you do what I'm doing now where I say, "well, the compiler did some inferences so whatever the types are they must be correct". Or maybe I'm thinking too hard about this and type annotations would make human consumption easier (though I'm still curious about the above questions).

Reply:
In safe code I tend to follow the same approach -- if the compiler found a type that makes it all work, then that seems right (but I will still sometimes add type annotations to make the code more readable).

In unsafe code though, I don't like trusting the compiler like that. Types are much less meaningful in unsafe code (that's why the code is unsafe to begin with), so IMO type inference should only be relied upon for absolutely obvious cases -- which this one is not.

Reply:
Good point, will do.